text/0000-maybe-dangling.md

```diff
@@ -0,0 +1,247 @@
+There we have a function that invokes a user-supplied `impl FnOnce` closure, roughly like this:
+```rust
+// Not showing all the `'lifetime` tracking, the point is that
+// this closure might live shorter than `thread`.
+fn thread(control: ..., closure: impl FnOnce() + 'lifetime) {
+    closure();
+    control.signal_done();
+    // A lot of time can pass here.
+}
+```
+The closure has a non-`'static` lifetime, meaning clients can capture references to on-stack data.
+The surrounding code ensure that `'lifetime` lasts at least until `signal_done` is triggered, which ensures that the closure never accesses dangling data.
+
+However, note that `thread` continues to run even after `signal_done`! Now consider what happens if the closure captures a reference of lifetime `'lifetime`:
+- The type of `closure` is a struct (the implicit unnameable closure type) with a `&'lifetime mut T` field.
+  References passed to a function must be live for the entire duration of the call.
+- The closure runs, `signal_done` runs.
+  Then -- potentially -- this thread gets scheduled away and the main thread runs, seeing the signal and returning to the user.
+    Now `'lifetime` ends and the memory the reference points to might be deallocated.
+- Now we have UB! The reference that as passed to `thread` with the promise of remaining live for the entire duration of the function, actually got deallocated while the function still runs. Oops.
+
+### Example 3
+
+As a third example, consider a type that wants to store a "pointer together with some data borrowed from that pointer", like the `owning_ref` crate. This will usually boil down to something like this:
+
+```rust
+unsafe trait StableDeref: Deref {}
+
+struct OwningRef<U, T: StableDeref<Target=U>> {
+    buffer: T,
+    ref_: NonNull<U>, // conceptually borrows from `buffer`.
+}
+```
+
+Such a type is unsound when `T` is `&mut U` or `Box<U>` because those types are assumed by the compiler to be unique, so any time `OwningRef` is passed around, the compiler can assume that `buffer` is a unique pointer -- an assumption that this code breaks because `ref_` points to the same memory!
+
+### Goal of this RFC
+
+The goal of this RFC is to
+- make the first example UB-free without code changes
+- make the second example UB-free without needing to add `unsafe` code
+- make it possible to define a type like the third example
+
+(Making the 2nd example UB-free without code changes would incur cost across the ecosystem, see the alternatives discussed below.)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+To handle situations like this, Rust has a special type called `MaybeDangling<P>`:
+references and boxes in `P` do *not* have to be dereferenceable or follow aliasing guarantees.
```

Comment:
question: what about transitive references? e.g. in `MaybeDangling<Box<Vec<T>>>`, is the inner vec also allowed to break aliasing guarantees?

Reply:
The only sensible interpretation is that it also applies to transitive references. If the box itself isn't necessarily a valid pointer then how could we even talk about the stuff it points to?

An alternative explanation of this behavior is that `MaybeDangling<T>` "truncates" the validity invariant of `T` to only the portion that does not involve the state of memory - only the bytes within `T` itself are relevant. So clearly memory pointing to more memory is also irrelevant.

Reply:
Yeah, I agree that that's the only way this feature is _useful_, but I'm not familiar enough with e.g. miri to know if that's easily implementable. Probably.

Reply:
> to know if that's easily implementable

Miri (and Stacked Borrows more generally) already does not care about the memory contents of a reference beyond the fact that it exists. Under the Stacked Borrows model, `MaybeDangling` only impacts the `Box` in `MaybeDangling<Box<Vec<T>>>`, but the `Vec` doesn't have any validity requirements enforced on it from passing around the `Box` anyway.

It's also completely impossible for the `Box` to not have a validity requirement to point to something real and also have a requirement on what the `Box` points to. If the pointed to thing mayn't exist, it can't be required to be well formed.

So `MaybeDangling` does remove all indirected-memory-based validity requirements of the type it wraps, including any transitive requirements that may exist. None do under Stacked Borrows, but could theoretically exist.

`MaybeDangling` as currently specified by the RFC does not change safety invariants, as it implements safe `Deref` conversion like `ManuallyDrop` does. I argue that this is likely undesirable, and `MaybeDangling` should instead make it `unsafe` to reactivate the wrapped type and its invariants.

Reply:
Might be worth explicitly mentioning this interaction in the RFC so it's clear that nested allocations in self-ref types will still work.

Reply:
The RFC already says this:

> Another way to think about this is: most types only have "by-value" requirements for their validity, i.e., they only require that the bit pattern be of a certain shape.
References and boxes are the sole exception, they also require some properties of the memory they point to (e.g., they need to be dereferenceable).
`MaybeDangling<T>` is a way to "truncate" `T` to its by-value invariant, which changes nothing for most types, but means that references and boxes are allowed as long as their bit patterns are fine (aligned and non-null) and as long as there *conceivably could be* a state of memory that makes them valid (`T` is inhabited).

That was meant to clarify exactly this point. Happy for suggestions for how to make this more clear!

Reply:
Ah, hm, that makes sense. And there are examples of self-ref stuff as well. Perhaps we should mention nested pointers in the examples too? I think that's why I missed it.


While RFCs aren't documentation, in general I think good unsafe abstraction docs:

 - spell out what is and isn't safe normally
 - spell out what the abstraction does in that context
 - have sufficient examples, explaining what the abstraction did and didn't do in those contexts (e.g. in this context: it did change how the pointer is treated, but it did not change anything about pointers-behind-pointers because they don't matter here)

i think that's been one of the biggest problems with the Pin docs (to be fixed by https://github.com/rust-lang/rust/pull/88500, but that's stale now) because it's incredibly unclear where the actual compiler safety magic lies, if any.

Not a big deal though, we can iterate on this whilst writing docs, and I'm happy to be involved then.